File: 0020-valid-parentheses/0020-valid-parentheses.py
# s - STRING

class Solution:
    def isValid(self, s: str) -> bool:
        # Use a stack to LIFO to ensure the opening is properly closed.
        stack = []
        
        openBracket = '([{'
        closeBracket = ')]}'
        
        for char in s:
            if char in closeBracket:
                # If closing character, pop stack and check if its the correspoding opening pair.
                
                if not stack:
                    # Return False if stack is empty while iterating.
                    return False

                # Peek into stack.
                top = stack[-1]

                # If the top of the stack is the corresponding opening pair, its correct so pop and continue.
                if top == openBracket[closeBracket.find(char)]:
                    stack.pop()
                else:
                    # If not, return False.
                    return False
            else:
                # If opening character, push to stack.
                stack.append(char)   

        # If stack empty after all iterations, means correct so return True.
        if not stack:
            return True

File: 0020-valid-parentheses/0020-valid-parentheses.java
import java.util.Stack;

class Solution {
    
    // Helper function to check for matching pairs.
    private boolean isMatching(char open, char close) {
        return (open == '(' && close == ')') ||
            (open == '{' && close == '}') ||
            (open == '[' && close == ']');
    }
    
    public boolean isValid(String s) {
        // Initialize a stack.
        Stack<Character> stack = new Stack<>();
        
        // Return false for uneven length strings.
        if ((s.length() % 2 != 0)) {
            return false;
        } 
            
        for (int i = 0; i < s.length(); i++) {
            // Get first character.
            char currChar = s.charAt(i);

            if ( currChar == ')' || currChar == '}' || currChar == ']') {
                if (stack.isEmpty() || !isMatching(stack.peek(), currChar)){
                    // If stack is empty or is not a matching pair, return false.
                    return false;
                } else {
                    // Pop it and continue.
                    stack.removeLast();
                }
            } else {
                // Push to stack if opening bracket.
                stack.add(currChar);
            }
        }
        // Return if stack is empty or not; all matched should be empty, otherwise it will not.
        return stack.isEmpty();
    }
}